### DISCUSSION TITLE
zigpy depends on rpds since 0.63.5

### DISCUSSION THREAD

**Discussion by pipiche38** [zigpy/zigpy#1375] [General]
Since a recent update introduced by https://github.com/zigpy/zigpy/blob/24d3a00fbbf952b96d1216547275a5cce4e4cf0e/zigpy/ota/providers.py#L19 in commit https://github.com/zigpy/zigpy/commit/a87b23606c851e145b99a3bc5589429db72551e1

the zigpy library from and above 0.63.5 is not compatible with Embeded Python Framework and so forth with Domoticz.

This issue was existing with _asyncio (a C implementation of asyncio) and a workaround was found. However in the case of rpds which is called by jsonschema via referencing nothing exist.

The problem has been reported to rpds.

> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     import zigpy.application
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/zigpy/application.py", line 22, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     import zigpy.appdb
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/zigpy/appdb.py", line 16, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     import zigpy.device
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/zigpy/device.py", line 12, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     from zigpy.ota.manager import find_ota_cluster, update_firmware
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/zigpy/ota/__init__.py", line 27, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     import zigpy.ota.providers
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/zigpy/ota/providers.py", line 19, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     import jsonschema
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/jsonschema/__init__.py", line 13, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     from jsonschema._format import FormatChecker
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/jsonschema/_format.py", line 11, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     from jsonschema.exceptions import FormatError
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/jsonschema/exceptions.py", line 16, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:     from referencing.exceptions import Unresolvable as _Unresolvable
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.357  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/referencing/__init__.py", line 4, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.358  Error: Zigpy-deConz:     from referencing._core import Anchor, Registry, Resource, Specification
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.358  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/referencing/_core.py", line 9, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.358  Error: Zigpy-deConz:     from rpds import HashTrieMap, HashTrieSet, List
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.358  Error: Zigpy-deConz:   File "/usr/local/lib/python3.10/site-packages/rpds/__init__.py", line 1, in <module>
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.358  Error: Zigpy-deConz:     from .rpds import *
> Apr 04 16:34:11 rasp domoticz[29854]: 2024-04-04 16:34:11.358  Error: Zigpy-deConz: ImportError: PyO3 modules do not yet support subinterpreters, see https://github.com/PyO3/pyo3/issues/576

**Comment by puddly** [comment]
We really don't try to ensure zigpy has only pure-Python dependencies. I've looked through the dependency tree and it looks like jsonschema's dependency `referencing` only switched to their own C library in a later version. You can downgrade it to an earlier one:

```
pip install referencing==0.23.0
```

Zigpy unit tests pass so you may be able to use it.

**Reply by pipiche38** [reply]
Thanks, will investigate that straight

**Reply by pipiche38** [reply]
Yes it does the trick, but looks like something is pushing to get the latest version :-(

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
jsonschema-specifications 2023.12.1 requires referencing>=0.31.0, but you have referencing 0.23.0 which is incompatible.
```

**Reply by pipiche38** [reply]
it looks like downgrading to jsonschema==4.17.3 is enought; is that good enough on zigpy side ?

```
jsonschema==4.17.3
jsonschema-specifications==2023.12.1
referencing==0.34.0
```

**Reply by puddly** [reply]
Zigpy doesn't pin a version for `jsonschema`:
https://github.com/zigpy/zigpy/blob/24d3a00fbbf952b96d1216547275a5cce4e4cf0e/pyproject.toml#L26

Create a constraint file for `pip` for your project and you should be fine: https://pip.pypa.io/en/stable/user_guide/#constraints-files
